packages/@aws-cdk/aws-cloudwatch/lib/env-tokens.ts

```diff
@@ -0,0 +1,75 @@
+import * as cdk from '@aws-cdk/core';
+import { dropUndefined } from './metric-util';
+
+/**
+ * Make a Token that renders to given region if used in a different stack, otherwise undefined
+ */
+export function regionIfDifferentFromStack(region: string): string {
+  return cdk.Token.asString(new StackDependentToken(region, stack => stack.region));
+}
+
+/**
+ * Make a Token that renders to given account if used in a different stack, otherwise undefined
+ */
+export function accountIfDifferentFromStack(account: string): string {
+  return cdk.Token.asString(new StackDependentToken(account, stack => stack.account));
+}
+
+/**
+ * A lazy token that requires an instance of Stack to evaluate
+ */
+class StackDependentToken implements cdk.IResolvable {
+  public readonly creationStack: string[];
+  constructor(private readonly originalValue: string, private readonly fn: (stack: cdk.Stack) => string) {
+    this.creationStack = cdk.captureStackTrace();
+  }
+
+  public resolve(context: cdk.IResolveContext) {
+    const stackValue = this.fn(cdk.Stack.of(context.scope));
+
+    if (cdk.Token.isUnresolved(stackValue) || stackValue === this.originalValue) {
+      return undefined;
+    }
+
+    return this.originalValue;
+  }
+
+  public toString() {
+    return cdk.Token.asString(this);
+  }
+
+  public toJSON() {
+    return this.originalValue;
+  }
+}
+
+/**
+ * A token that will not return an empty "options" object as the last element in an array
+ *
+ * Necessary to prevent options objects that only contain "region" and "account" keys
+ * that evaluate to "undefined" from showing up in the rendered JSON.
+ */
+export class DropEmptyGraphOptions implements cdk.IResolvable, cdk.IPostProcessor {
```

Comment:
> There's nothing in this class that is specific to graph options, is there?

There is. It knows to interpret its argument as an array which optionally can have an object as its last element, and we operate on the last element. That pattern is only used in CloudWatch graphs.

> while `dropUndefined` is part of `metric-util`

`dropUndefined` is more generic and also used in other places. I can rename this file, but the behavior in this class is so... "special" (given it uses Token magic and all) that I wanted to keep it separate. It's called `env-tokens`, because it makes Tokens that are environment-related.

> I'd say we should avoid util classes or files

I don't really agree. I make utility functions for functionality that *should* have been in a JavaScript's standard library all the time (see `dropUndefined`) and I tend to put them in a file `util.ts` or similar. The alternatives to such a util file are:

* taking a dependency on `underscore` (please god no)
* copy/pasting the implementation of `dropUndefined` into every file that needs it (also, please god no? but ymmv)

I agree that I mixed up two different classes of utilities in the current `metric-util`, which is bad on my part. I will split it out into a proper `util` and something about metric expansion.

Reply:
>> There's nothing in this class that is specific to graph options, is there?

> There is. It knows to interpret its argument as an array which optionally can have an object as its last element, and we operate on the last element. That pattern is only used in CloudWatch graphs.

We should be naming classes based on what it does and what it operates on; not where it is used - that's what namespaces are for (not typescript namespaces, but conceptual) and it's already correctly placed under `aws-cloudwatch`.

Further, it's not an `env-token` so that's not the right spot either. I still think it should be called something like `DropLastEmptyElement` and place it under a similarly named file?